### PR TITLE
flip camera2D init and from_raw_data

### PR DESCRIPTION
**Pull request by alejcas**
This is the camera init/from_raw_data code flip we talk on discord.

Nota that from_raw_data cameras won't reference the same data objects but create new ones using data from the provided objects.

I have also fixed examples and tests.

I tested using the example codes and all worked

### CODE REVIEW

**Comment by DragonMoffon** [comment]
While I understand why it was done the way it was, this fundamentally breaks a core assumption of the Cameras. The CameraData and OrthographicProjectionData objects have to be shared between objects.
It would be better to create the camera with default values and then override the attributes later

**Comment by alejcas** [comment]
I can rewrite from_raw_data to build a camera with the arguments (c data and p data).

~~but i think it should be called “from_camera”:~~

~~`Camera2D.from_camera(camera: Camera2D)`~~

or “from_camera_data” with same arguments as “from_raw_data” but reusing the internal data objects.

~~what do you think?~~

rewrite done on fb025ec

File: arcade/camera/camera_2d.py [hunk 1]
```diff
@@ -49,104 +49,40 @@ class Camera2D:
...
     contents are exposed via properties rather than directly to prevent this.
 
-    :param window: The Arcade Window to bind the camera to.
-            Defaults to the currently active window.
-    :param camera_data: A :py:class:`~arcade.camera.data.CameraData`
-        describing the position, up, forward and zoom.
-    :param projection_data: A :py:class:`~arcade.camera.data.OrthographicProjectionData`
-        which describes the left, right, top, bottom, far, near planes and the viewport
-        for an orthographic projection.
+    :param viewport: A 4-int tuple which defines the pixel bounds which the camera with project to.
```
Comment (MiCurry) [inline arcade/camera/camera_2d.py]: 'with project to.' -> 'will project to'. Or 'will project onto'.

Comment (alejcas) [reply arcade/camera/camera_2d.py]: I didn't write that but it should be "will project to" I'll change it

File: arcade/camera/camera_2d.py [hunk 2]
```diff
@@ -49,104 +49,40 @@ class Camera2D:
...
+    :param position: The 2D position of the camera in the XY plane.
+    :param up: The 2D unit vector which defines the +Y-axis of the camera space.
```
Comment (MiCurry) [inline arcade/camera/camera_2d.py]: I like the description used in `CameraData` better here:

``` Python
        up: A 3D vector which describes which direction is up (+y).
```

At least for my brain, I like seeing the word 'up' (with the additional mention of the positive y direction). It helps me understand that this controls the normal vector of the camera.

File: arcade/camera/camera_2d.py [hunk 3]
```diff
@@ -49,104 +49,40 @@ class Camera2D:
...
+            i.e. a zoom of 2.0 halves the size of the projection, doubling the perceived size of objects.
+    :param projection: A 4-float tuple which defines the world space
+                bounds which the camera projects to the viewport.
+    :param near: The near clipping plane of the camera.
+    :param far: The far clipping plane of the camera.
     :param render_target: The FrameBuffer that the camera uses. Defaults to the screen.
         If the framebuffer is not the default screen nothing drawn after this camera is used will
         show up. The FrameBuffer's internal viewport is ignored.
+    :param window: The Arcade Window to bind the camera to.
+        Defaults to the currently active window.
```
Comment (MiCurry) [inline arcade/camera/camera_2d.py:65]: Should we keep these comments written the same as the comments in `CameraData` and `OrthographicProjectionData`? 🤷‍♂️ Just something to think about.

File: arcade/camera/camera_2d.py [hunk 4]
```diff
@@ -165,26 +101,75 @@ def from_raw_data(
...
+        )
+
+    @classmethod
+    def from_camera_data(cls, *,
+                         camera_data: Optional[CameraData] = None,
+                         projection_data: Optional[OrthographicProjectionData] = None,
+                         render_target: Optional[Framebuffer] = None,
+                         window: Optional["Window"] = None) -> Self:
+        """
+        Create a Camera2D with reusing provided CameraData or/and OrthographicProjectionData object.
```
Comment (MiCurry) [inline arcade/camera/camera_2d.py]: How about:
```
Create a Camera2D from provided CameraData or/and OrthographicProjectionData objects.
```

Because it's possible that someone may use this function without another Camera. They may just create their own `CameraData` and/or `OrthographicProjectionData` objects.

File: arcade/camera/camera_2d.py [hunk 5]
```diff
@@ -165,26 +101,75 @@ def from_raw_data(
...
+                    f"projection height is 0 due to equal {bottom=}"
+                    f"and {top=}"))
+            near, far = projection_data.near, projection_data.far
+            if near == far:
+                raise ZeroProjectionDimension(
+                    f"projection depth is 0 due to equal {near=}"
+                    f"and {far=} values"
+                )
+
+        # build a new camera with defaults and after reuse provided camera objects
```
Comment (MiCurry) [inline arcade/camera/camera_2d.py]: How about: 

```
 build a new camera with defaults and then apply the provided camera objects.
 ```

**Comment by MiCurry** [comment]
I'm wondering if we might run into some shallow vs deep copy issues with `from_camera_data`. Specifically if someone tries to use the camera data from one camera to initialize another:

``` python
camera1 = Camera(...)
camera2 = Camera.from_camera_data(camera1.view_data, camera1.projection_data)
```

In that case I think adjusting either camera would affect the other, since they both contain references to each other. We should be able to easily test this in a unit test, which I'll make if I get some freetime today.

File: arcade/camera/camera_2d.py (hunk 2 above)
Comment (alejcas) [reply arcade/camera/camera_2d.py]: But in camera 2D it's a 2D vector...
I'll combine both with 2d

File: arcade/camera/camera_2d.py (hunk 3 above)
Comment (alejcas) [reply arcade/camera/camera_2d.py:65]: Yes, they should be there. A normal user will use a camera2D and not know nothing about what is CameraData.

File: arcade/camera/camera_2d.py (hunk 4 above)
Comment (alejcas) [reply arcade/camera/camera_2d.py]: Ok

File: arcade/camera/camera_2d.py (hunk 5 above)
Comment (alejcas) [reply arcade/camera/camera_2d.py]: ok

**Comment by alejcas** [comment]
> I'm wondering if we might run into some shallow vs deep copy issues with `from_camera_data`. Specifically if someone tries to use the camera data from one camera to initialize another:
> 
> ```python
> camera1 = Camera(...)
> camera2 = Camera.from_camera_data(camera1.view_data, camera1.projection_data)
> ```
> 
> In that case I think adjusting either camera would affect the other, since they both contain references to each other. We should be able to easily test this in a unit test, which I'll make if I get some freetime today.

I think this is the intended purpose.

File: arcade/camera/camera_2d.py [hunk 6]
```diff
@@ -165,26 +101,75 @@ def from_raw_data(
...
+        )
+
+    @classmethod
+    def from_camera_data(cls, *,
+                         camera_data: Optional[CameraData] = None,
+                         projection_data: Optional[OrthographicProjectionData] = None,
+                         render_target: Optional[Framebuffer] = None,
+                         window: Optional["Window"] = None) -> Self:
+        """
+        Create a Camera2D from provided CameraData or/and OrthographicProjectionData objects.
```
Comment (pushfoo) [inline arcade/camera/camera_2d.py]: TL;DR: Something like this for explaining the purpose of the function?
```suggestion
        """Make a ``Camera2D`` directly from data objects.

        This :py:class:`classmethod` allows advanced users to:

        #. skip or replace the default validation
        #. share ``camera_data`` or ``projection_data`` between cameras

       .. warning:: Be careful when sharing data objects!

                    **Any** action on a camera which changes a shared
                    object changes it for **every** camera which uses
                    the same object.

       .. list-table::
          :header-rows: 1

          * - Shared Value
            - Example Use(s)

          * - ``camera_data``
            - Mini-maps, reflection, and ghosting effects.

          * - ``projection_data``
            - Simplified rendering configuration

          * - ``render_target``
            - Complex rendering setups

```

**Review by pushfoo — CHANGES_REQUESTED** [review]
**TL;DR: The docstring change to clarify the shared data object behavior is the most we should do in this PR.**

@MiCurry the shared behavior is intended, but I agree we need to document it better and think about the CameraAPI a bit more.

However, this PR isn't the place for those changes. We can add a minimal docstring fix since @DragonMoffon and @DigiDuncan are finishing up prototyping improvements which will likely cause merge conflicts with further API changes.

File: arcade/camera/camera_2d.py (hunk 6 above)
Comment (alejcas) [reply arcade/camera/camera_2d.py]: I don't understand the dots before the warning or the list-table, and also the alignment of those docstrings.
I'll submit the change anyway in a few moments

Comment (MiCurry) [reply arcade/camera/camera_2d.py]: @pushfoo yeah that looks good to me!

**Review by pushfoo — APPROVED** [review]
LGTM. A bunch of changes are coming, so we'll fix lingering API & doc issues in subsequent PRs.
